drivers/ipm/ipm_rpi_pico.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2025 Dan Collins
+ * Copyright (c) 2025 Dmitrii Sharshakov <d3dx12.xx@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_sio_fifo
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/ipm.h>
+
+#include <zephyr/logging/log.h>
+#include "hardware/structs/sio.h"
+
+LOG_MODULE_REGISTER(ipm_rpi_pico, CONFIG_IPM_LOG_LEVEL);
+
+struct rpi_pico_mailbox_config {
+	sio_hw_t *const sio_regs;
+};
+
+struct rpi_pico_ipm_data {
+	ipm_callback_t cb;
+	void *user_data;
+};
+
+static struct rpi_pico_mailbox_config rpi_pico_sio_mailbox_config = {
+	.sio_regs = (sio_hw_t *)DT_INST_REG_ADDR_BY_NAME(0, sio),
+};
+static struct rpi_pico_ipm_data rpi_pico_mailbox_data;
+
+static int rpi_pico_mailbox_send(const struct device *dev, int wait, uint32_t id,
+	const void *data, int size)
+{
+	ARG_UNUSED(data);
+
+	if (size != 0) {
+		return -EMSGSIZE;
+	}
+
+	if (id > UINT32_MAX) {
+		return -EINVAL;
+	}
+
+	const struct rpi_pico_mailbox_config *config =
+		(const struct rpi_pico_mailbox_config *)dev->config;
+
+	if (!(config->sio_regs->fifo_st & SIO_FIFO_ST_RDY_BITS) && !wait) {
+		LOG_ERR("Mailbox FIFO is full, cannot send message");
+
+		return -EBUSY;
+	}
+
+	while (!(config->sio_regs->fifo_st & SIO_FIFO_ST_RDY_BITS)) {
+	}
+
+	config->sio_regs->fifo_wr = id;
+
+	/* Inform other CPU about FIFO update. */
+	__SEV();
+
+	return 0;
+}
+
+static void rpi_pico_mailbox_register_callback(const struct device *dev, ipm_callback_t cb,
+					       void *user_data)
+{
+	struct rpi_pico_ipm_data *data = dev->data;
+
+	data->cb = cb;
+	data->user_data = user_data;
+}
+
+static int rpi_pico_mailbox_max_data_size_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	/*
+	 * FIFO mailbox allows a single 32 bit value to be sent - and we
+	 * use that as the channel identifier.
+	 */
+	return 0;
```

Comment:
Could it maybe make sense to do some kind of a split, like 1/3 or 2/2 split for ID/Data? 2^32 seems like quite a huge number of channels for most applications, while sending some data as well might be a cool feature for debugging or if any IPM/MBOX consumers want to transfer some data as well